src/modules/ekf2/ekf2_main.cpp

```diff
@@ -1663,73 +1692,153 @@ void Ekf2::run()
-					float vel_ne_innov_length = sqrtf(innovations.vel_pos_innov[0] * innovations.vel_pos_innov[0] +
-									  innovations.vel_pos_innov[1] * innovations.vel_pos_innov[1]);
-					_preflt_horiz_fail = (_vel_ne_innov_lpf.norm() > _vel_innov_test_lim)
-							     || (vel_ne_innov_length > 2.0f * _vel_innov_test_lim)
-							     || (_yaw_innov_magnitude_lpf > yaw_test_limit);
+	// master pass-fail status
+	_preflt_fail = _preflt_horiz_fail || _preflt_vert_fail;
+}
 
-					// check the vertical velocity and position innovations
-					_preflt_vert_fail = (fabsf(_vel_d_innov_lpf) > _vel_innov_test_lim)
-							    || (fabsf(innovations.vel_pos_innov[2]) > 2.0f * _vel_innov_test_lim)
-							    || (fabsf(_hgt_innov_lpf) > _hgt_innov_test_lim);
+bool Ekf2::preFlightCheckHeadingFailed(const filter_control_status_u &control_status,
+				       const vehicle_status_s &vehicle_status,
+				       const ekf2_innovations_s &innov,
+				       float alpha)
 
-					// master pass-fail status
-					_preflt_fail = _preflt_horiz_fail || _preflt_vert_fail;
+{
+	// Get the correct yaw test limit value depending on vehicle type and aiding mode
+	float yaw_test_limit = selectYawTestLimit(control_status, vehicle_status);
+	_filter_yaw_magnitude_innov.setSpikeLimit(2.0f * yaw_test_limit);
 
-				} else {
-					_vel_ne_innov_lpf.zero();
-					_vel_d_innov_lpf = 0.0f;
-					_hgt_innov_lpf = 0.0f;
-					_preflt_horiz_fail = false;
-					_preflt_vert_fail = false;
-					_preflt_fail = false;
-				}
+	float heading_innov_lpf = _filter_yaw_magnitude_innov.update(innov.heading_innov, alpha);
 
-				_estimator_innovations_pub.publish(innovations);
-			}
-		}
+	return checkInnovFailed(innov.heading_innov, heading_innov_lpf, yaw_test_limit);
+}
 
-		// publish ekf2_timestamps
-		_ekf2_timestamps_pub.publish(ekf2_timestamps);
+float Ekf2::selectYawTestLimit(const filter_control_status_u &control_status, const vehicle_status_s &vehicle_status)
+{
+	// set the max allowed yaw innovaton depending on whether we are not aiding navigation using
+	// observations in the NE reference frame.
+	bool doing_ne_aiding = control_status.flags.gps ||  control_status.flags.ev_pos;
+
+	float yaw_test_limit;
+
+	if (doing_ne_aiding && vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
+		// use a smaller tolerance when doing NE inertial frame aiding as a rotary wing
+		// vehicle which cannot use GPS course to realign heading in flight
+		yaw_test_limit = _nav_yaw_innov_test_lim;
+
+	} else {
+		// use a larger tolerance when not doing NE inertial frame aiding or
+		// if a fixed wing vehicle which can realign heading using GPS course
+		yaw_test_limit = _yaw_innov_test_lim;
+	}
+
+	return yaw_test_limit;
+}
+
+bool Ekf2::preFlightCheckHorizVelFailed(const filter_control_status_u &control_status,
+					const ekf2_innovations_s &innov,
+					float alpha)
+{
+	bool has_failed = false;
+
+	bool doing_ne_aiding = control_status.flags.gps ||  control_status.flags.ev_pos;
```

Comment:
```suggestion
	bool doing_hpos_aiding = control_status.flags.gps ||  control_status.flags.ev_pos;
```

Reply:
@mhkabir Since you want to rename that hpos because EV fusion doesn't mean NE navigation, it means that you don't actually really care of heading. Should we then use the test limit in the `else` part of the selection?